cranelift/codegen/src/isa/x64/inst/emit.rs

```diff
@@ -0,0 +1,2987 @@
+    let prefix66 = (flags & F_PREFIX_66) != 0;
+    let clearRexW = (flags & F_CLEAR_REX_W) != 0;
+    let retainRedundant = (flags & F_RETAIN_REDUNDANT_REX) != 0;
+    // The operand-size override, if requested.  This indicates a 16-bit
+    // operation.
+    if prefix66 {
+        sink.put1(0x66);
+    }
+    match memE {
+        Addr::IR { simm32, base: regE } => {
+            // First, cook up the REX byte.  This is easy.
+            let encE = iregEnc(*regE);
+            let w = if clearRexW { 0 } else { 1 };
+            let r = (encG >> 3) & 1;
+            let x = 0;
+            let b = (encE >> 3) & 1;
+            let rex = 0x40 | (w << 3) | (r << 2) | (x << 1) | b;
+            if rex != 0x40 || retainRedundant {
+                sink.put1(rex);
+            }
+            // Now the opcode(s).  These include any other prefixes the caller
+            // hands to us.
+            while numOpcodes > 0 {
+                numOpcodes -= 1;
+                sink.put1(((opcodes >> (numOpcodes << 3)) & 0xFF) as u8);
+            }
+            // Now the mod/rm and associated immediates.  This is
+            // significantly complicated due to the multiple special cases.
+            if *simm32 == 0
+                && encE != regs::ENC_RSP
+                && encE != regs::ENC_RBP
+                && encE != regs::ENC_R12
+                && encE != regs::ENC_R13
+            {
+                // FIXME JRS 2020Feb11: those four tests can surely be
+                // replaced by a single mask-and-compare check.  We should do
+                // that because this routine is likely to be hot.
+                sink.put1(mkModRegRM(0, encG & 7, encE & 7));
+            } else if *simm32 == 0 && (encE == regs::ENC_RSP || encE == regs::ENC_R12) {
+                sink.put1(mkModRegRM(0, encG & 7, 4));
+                sink.put1(0x24);
+            } else if low8willSXto32(*simm32) && encE != regs::ENC_RSP && encE != regs::ENC_R12 {
+                sink.put1(mkModRegRM(1, encG & 7, encE & 7));
+                sink.put1((simm32 & 0xFF) as u8);
+            } else if encE != regs::ENC_RSP && encE != regs::ENC_R12 {
+                sink.put1(mkModRegRM(2, encG & 7, encE & 7));
+                sink.put4(*simm32);
+            } else if (encE == regs::ENC_RSP || encE == regs::ENC_R12) && low8willSXto32(*simm32) {
+                // REX.B distinguishes RSP from R12
+                sink.put1(mkModRegRM(1, encG & 7, 4));
+                sink.put1(0x24);
+                sink.put1((simm32 & 0xFF) as u8);
+            } else if encE == regs::ENC_R12 || encE == regs::ENC_RSP {
+                //.. wait for test case for RSP case
+                // REX.B distinguishes RSP from R12
+                sink.put1(mkModRegRM(2, encG & 7, 4));
+                sink.put1(0x24);
+                sink.put4(*simm32);
+            } else {
+                panic!("emit_REX_OPCODES_MODRM_SIB_IMM_encG_memE: IR");
```

Comment:
```suggestion
                sink.put1(mkModRegRM(1, encG & 7, 5)); // could/?should? be encE & 7 for consistency
                sink.put1(0x25);
                sink.put1((simm32 & 0xFF) as u8);
```
and similar for `!low8WillSXto32()` would encode the `RBP`/`R13` variants that are currently missing here right? is there a wider reason this is `panic!`, or just "not gotten to it yet"?

Reply:
Hmm, I thought I had exhaustively tested this in the encoding test cases (there are 100+ amode encoding tests).  So I'm not convinced there's a missing case (am not convinced there isn't a missing case, either).  It's simple enough to shake out, by checking the encoding test cases.

Reply:
Ah, you're right - the initial checks against `R{BP,13}` had me looking for an else mentioning them, but they're encoded through the `if low8willSXto32(*simm32) && encE != regs::ENC_RSP && encE != regs::ENC_R12` case. There are a few tests for these such as for [`127(%rbp)`](https://github.com/bytecodealliance/wasmtime/pull/1605/files/f44e562b0660ed38ec8f01c94bf07efe3d4fe5fc#diff-4543aa3149dc21a011981c094c49644bR987-R991).

How do you feel about upgrading this to an `unreachable!`?

Reply:
I would be the first to admit that the logic is pretty hard to follow there.  Er, if `unreachable!` is the right thing there, then use that.  Just so long as it really *is* unreachable.